File: 0994-rotting-oranges/0994-rotting-oranges.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var orangesRotting = function(grid) {
   /*
   bfs
   1. init values
   2. check edge cases
   3. traverse 2d arrs to find rotten orange and fresh oranges. 
    if find rotten, put its position into queue. 
    if find fresh, increment them
   4. while queue exists, get curVal from queue.shift()
    5. get queue's size and traverse surrouding areas to find rottens
    6. only visit surrounding areas if there is a fresh oranges\
        7. change cur orange to rotten
        7. push cur position into queue
        8. fresh--
    9. after all visit curQue' size, mins++
   */
    
    let mins = 0;
    let row = grid.length;
    let col = grid[0].length;
    let queue = [];
    let fresh = 0;
    
    for(let i=0; i<row; i++){
        for(let j=0; j<col; j++){
            if(grid[i][j] === 2){
                queue.push([i,j])
            } else if(grid[i][j] === 1) fresh++;
        }
    }
    
    while(queue.length){
      
        let size = queue.length;
        for(let i=0; i<size; i++){
            let [x,y] = queue.shift();
            if(x-1 >=0 && grid[x-1][y] === 1 ){
                queue.push([x-1,y]);
                fresh--;
                grid[x-1][y] = 2;
            }
            if(x+1 < row && grid[x+1][y] === 1 ){
                queue.push([x+1,y]);
                fresh--;
                grid[x+1][y] = 2;
            }
            if(y-1 >=0 && grid[x][y-1] === 1 ){
                queue.push([x,y-1]);
                fresh--;
                grid[x][y-1] = 2;
            }
            if(y+1 < col && grid[x][y+1] === 1 ){
                queue.push([x,y+1]);
                fresh--;
                grid[x][y+1] = 2;
            }
        }
        if(queue.length) mins++;
    }
    return fresh === 0 ? mins : -1
};

File: 0994-rotting-oranges/0994-rotting-oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        '''
        U:
           q) can grid be empty?
           q) initially rotten oranges are located in any place?
           key points: if a fresh orange keeps remaining, return -1
           
           ex) grid = [[2,1,1],
                       [0,1,1],
                       [1,0,1]]
                       
        M: BFS
        
        P:
            1. init a queue, set and time val
            2. traverse grid to find rotten and fresh oranges
            2. while queue and set exist
                2.1) traverse around the cur position with looping
                    2.2) pop cur row and col from que
                    2.3) visit around the cur position to find whether there is a fresh orange, if it is, it gets rotten (put into queue and set)
                2.4) time ++
            3. iterative the grid to find fresh oranges, if find, return -1. 
            4. return time
            
        time: o(nm) nm is the size of the grid. space: o(nm) n is the size of the grid, m is for the queue
        
        
        P:
            1. init a queue, time val and set
            2. traverse the input grid to find rotten and fresh oranges. if rotten oranges are found, put into queue. elif fresh is found, put into set
            3. while que exists
                3.1) visit surround of cur position by looping
                    3.2) cur position is popped from queue  
                    3.3) using a loop, visit 4 directions from the cur position
                        3.4) if a fresh orange is found within valid boundaries, remove the freshorange from set and put cur position into queue
                4.) time++
            5. return -1 if set exist otherwise time
        '''
       
        que = deque()
        visit = set()
        time = 0
        rows , cols = len(grid), len(grid[0])
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 2:
                    que.append((i,j))
                elif grid[i][j] == 1:
                    visit.add((i,j))
                    
        while que and visit:
            for _ in range(len(que)):
                r,c = que.popleft()
                
                for dr, dc in [(1,0), (-1,0), (0,1), (0,-1)]:
                    row , col = r+dr, c+dc
                    
                    if 0<=row<rows and 0<=col<cols and (row, col) in visit:
                        visit.remove((row, col))
                        que.append((row, col))
            time+=1
            
        return -1 if visit else time
        